### PR TITLE
Refactor rewrite logging #375

### PR DESCRIPTION
**Pull request by JohannesPeter1984**

### CODE REVIEW

**Review by renekrie — CHANGES_REQUESTED** [review]
Thank you, @JohannesPeter1984 !

Not a full review yet but maybe you could already have a look at the package structure and some class names and maybe change it:

`querqy.model.logging`: `querqy.model` contains the classes of Querqy's query object model. It doesn't make sense to me to put logging underneath it. Maybe you can just pull 'logging' one level up? Alternatively, it could also be moved under `querqy.rewrite`.

You also created a new package `querqy.rewriter` with 2 `*Output` classes in it. I think I'd just move them to `querqy.rewrite` and remove `querqy.rewriter`. (I get it that they are a kind of model classes to catch different outputs but that would just be a very small model and it would not cover everything involved to represent the result of a rewriting - I don't find them worth a separate package).

Coming back to the classes in `querqy.model.logging`: With the exception of `RewriteLoggingConfig` they all represent classes that will be serialised as log entries. I think it would be fine to move `RewriteLoggingConfig` out of this package for clarity (maybe to `querqy.rewrite`). I also found the class names confusing: `ActionLogging` - sounds like a logging component but in the end it just represents a log message or log entry. Maybe `ActionLogEntry` ?

I also left some first comments.

File: querqy-core/src/main/java/querqy/rewrite/commonrules/LineParser.java [hunk 1]
```diff
@@ -219,7 +220,7 @@ public static Object parse(final String line, final Input inputPattern,
                     // should never happen
                     return new ValidationError("Cannot parse line: " + line);
                 } else {
-                    return new SynonymInstruction(synonymTerms, boost);
+                    return new SynonymInstruction(synonymTerms, boost, InstructionDescription.builder().typeName("").build());
```
Comment (renekrie) [inline querqy-core/src/main/java/querqy/rewrite/commonrules/LineParser.java:223]: Use InstructionDescription.empty()

File: querqy-core/src/main/java/querqy/model/rewriting/RewriterOutput.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    public Optional<RewriterLogging> getRewriterLogging() {
+        return Optional.ofNullable(rewriterLogging);
+    }
+
+    public static RewriterOutputBuilder builder() {
+        return new RewriterOutputBuilder();
+    }
+
+    public static class RewriterOutputBuilder {
```
[the remark this answers is not in the document]
Comment (JohannesPeter1984) [reply querqy-core/src/main/java/querqy/model/rewriting/RewriterOutput.java:32]: As long as we use a builder we can enhance the output without any refactoring regarding existing tests (I want to avoid the need to refactor tons of tests just to let them additionally pass null). If we have multiple constructors, it is not obvious, which one is valid for regular use and which ones should only be used for tests. With a builder, this is very simple: You need to pass everything as long they are not defined as Optional<> in the underlying context object. For pure testing purposes, you can leave them out. 

This is why I would avoid using constructors whenever possible.

File: querqy-core/src/main/java/querqy/rewrite/RewriteChain.java [hunk 3]
```diff
@@ -45,41 +45,95 @@ public RewriteChain(final List<RewriterFactory> factories) {
...
+    public RewriteChainOutput rewrite(final ExpandedQuery query,
+                                      final SearchEngineRequestAdapter searchEngineRequestAdapter) {
 
-        final Optional<InfoLoggingContext> loggingContext = searchEngineRequestAdapter.getInfoLoggingContext();
+        final RewritingExecutor executor = new RewritingExecutor(factories, searchEngineRequestAdapter, query);
+        return executor.rewrite();
+    }
+
+    @Deprecated
+    public List<RewriterFactory> getRewriterFactories() {
```
Comment (renekrie) [inline querqy-core/src/main/java/querqy/rewrite/RewriteChain.java:56]: As far as I can see, this method is not used (same like the other Deprecated getFactory(1x)). Can we just remove it?

File: querqy-core/src/main/java/querqy/rewrite/commonrules/model/FilterInstruction.java [hunk 4]
```diff
@@ -21,27 +21,37 @@
...
-      }
-      this.filterQuery = filterQuery instanceof BooleanQuery
-              ? InstructionHelper.applyMinShouldMatchAndGeneratedToBooleanQuery((BooleanQuery) filterQuery)
-              : filterQuery;
-   }
-
-   /* (non-Javadoc)
-    * @see querqy.rewrite.commonrules.model.Instruction#apply(querqy.rewrite.commonrules.model.PositionSequence, querqy.rewrite.commonrules.model.TermMatches, int, int, querqy.model.ExpandedQuery, java.util.Map)
+    @Deprecated // Use only for testing
+    public FilterInstruction(final QuerqyQuery<?> filterQuery) {
```
Comment (renekrie) [inline querqy-core/src/main/java/querqy/rewrite/commonrules/model/FilterInstruction.java:27]: Let's make it package visible then or remove it and change the tests

File: querqy-core/src/main/java/querqy/rewrite/contrib/replace/ReplaceInstruction.java [hunk 5]
```diff
@@ -42,16 +43,37 @@ public void apply(final List<CharSequence> seq, final int start, final int exclu
      * @param start            Startposition of the term in the list
      * @param exclusiveOffset  Endposition of the term in the list without offset.
      * @param replacementTerms Terms that should be used as replacement
-     * @param appliedRules     Debug information about replaced terms and their replacement
+     * @param actionLoggings   Debug information about replaced terms and their replacement
```
Comment (renekrie) [inline querqy-core/src/main/java/querqy/rewrite/contrib/replace/ReplaceInstruction.java:46]: matchType is missing in Javadoc `@param` annotations - the javadoc generator will not accept this. Please fix this to make the release a bit easier ;-)

**Review by renekrie — CHANGES_REQUESTED** [review]
Thank you - just some minor things to check

File: querqy-core/src/main/java/querqy/rewrite/commonrules/model/FilterInstruction.java (hunk 4 above)
Comment (JohannesPeter1984) [reply querqy-core/src/main/java/querqy/rewrite/commonrules/model/FilterInstruction.java:27]: I cannot make them package private as they are used by many different test classes & code (e.g. deprecated LineParser)

Comment (JohannesPeter1984) [reply querqy-core/src/main/java/querqy/rewrite/commonrules/model/FilterInstruction.java:27]: (affects more or less all Instruction classes)

Comment (JohannesPeter1984) [reply querqy-core/src/main/java/querqy/rewrite/commonrules/model/FilterInstruction.java:27]: We anyway know that we will change the structure of Instruction sooner or later

File: querqy-core/src/main/java/querqy/model/rewriting/RewriterOutput.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (JohannesPeter1984) [reply querqy-core/src/main/java/querqy/model/rewriting/RewriterOutput.java]: Still the builder makes you more flexible in refactoring the structure of the context object (which is definitely going to happen for decorations in the future). I think there are valid arguments on both sides, so the decision is somehow a matter of personal preference.

File: querqy-core/src/main/java/querqy/rewrite/RewriteChain.java [hunk 6]
```diff
@@ -24,62 +23,112 @@ public class RewriteChain {
...
-    public List<RewriterFactory> getRewriterFactories() {
-        return factories;
-    }
+        private void addLoggingIfRewritingHasBeenApplied(final String factoryId, final RewriterLog rewriterLogging) {
+            if (rewriterLogging.hasAppliedRewriting()) {
+                addLoggingIfRewriterIdIsIncluded(factoryId, rewriterLogging);
+            }
+        }
+
+        private void addLoggingIfRewriterIdIsIncluded(final String factoryId, final RewriterLog rewriterLogging) {
```
Comment (renekrie) [inline querqy-core/src/main/java/querqy/rewrite/RewriteChain.java:110]: I'd prefer to not interpret includedRewriters.empty()=true as 'logging enabled for all rewriters. As soon as someone turns on one rewriter, suddenly all others become disabled. In ES we required each rewriter that is to be logged to have logging enabled - that should be the most robust option. I'd also prefer renaming `included` to something `loggingEnabled` (see querqy.elasticsearch.RewriteChainAndLogging#rewritersEnabledForLogging)

**Review by renekrie — CHANGES_REQUESTED** [review]
Thank you for your most recent updates. 
This is hopefully my final nit-picking!

File: querqy-core/src/main/java/querqy/rewrite/RewriteChain.java (hunk 6 above)
Comment (JohannesPeter1984) [reply querqy-core/src/main/java/querqy/rewrite/RewriteChain.java:110]: I think it is quite counterintuitive to set the logging as enabled and not to get anything back. Furthermore, I think it is not counterintuitive to return logging only for one rewriter if only one rewriter is specified. I do not see the use case of having logging enabled while activating no rewriter.

Comment (renekrie) [reply querqy-core/src/main/java/querqy/rewrite/RewriteChain.java:110]: Following our conversation, we will tackle this when we refactor the Sink concept in a follow-up step

**Review by renekrie — APPROVED** [review]
